components/Init/src/virtio_vsock.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+/**
+ * @brief virtio vsock backend driver layer for x86
+ *
+ * The virtio vsock backend has two layers, the emul layer and the driver layer.
+ * This file is part of the driver layer, it's responsible for:
+ *
+ * - TX: receiving data from the emul layer and forwarding them to the destination
+ *   VM using camkes virtqueue (which is different from the virtqueue that's
+ *   used to communicate between the guest and the emul layer)
+ * - RX: receiving data from the source port via camkes virtqueue and invoking the
+ *   emul layer to handle the data
+ * - virtio vsock backend initialization
+ * - sending virtual IRQ to the guest
+ *
+ * This file relies on the connection topology and camkes virtqueue connection
+ * being configured correctly in camkes.
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_vsock.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include <virtio/virtio_vsock.h>
+
+#include "vm.h"
+#include "virtio_vsock.h"
+#include "virtio_irq.h"
+
+static virtio_vsock_t *virtio_vsock;
+
+/* Used as the parameter for callback functions (currently only `vsock_inject_irq`) */
+typedef struct virtio_vsock_cookie {
+    vm_t *vm;
+} virtio_vsock_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(socket_layout)
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct conn {
+    int cid;
+    virtqueue_driver_t *send_queue;
+    virtqueue_device_t *recv_queue;
+} conn_t;
+static conn_t connections[NUM_PORTS];
+
+/**
+ * Cleans up and frees camkes buffers we allocated.
+ *
+ * @param queue Send queue
+ */
+static void virtio_vsock_notify_free_send(virtqueue_driver_t *queue)
+{
+    void *buf = NULL;
+    uint32_t buf_size = 0, wr_len = 0;
+    vq_flags_t flag;
+    virtqueue_ring_object_t handle;
+
+    while (virtqueue_get_used_buf(queue, &handle, &wr_len)) {
+        while (camkes_virtqueue_driver_gather_buffer(queue, &handle, &buf, &buf_size, &flag) >= 0) {
+            camkes_virtqueue_buffer_free(queue, buf);
+        }
+    }
+}
+
+/* Should be defined in virtio_vsock_emul.c */
+void WEAK vsock_rx_complete(virtio_emul_t *emul, char *buf, unsigned len)
+{
+    assert(!"should not be here");
+}
+
+/**
+ * Receives data from the source VM using camkes virtqueue, and invokes the emul layer
+ * to transmit the data to the guest VM.
+ *
+ * @param queue Receive queue
+ */
+static void virtio_vsock_notify_recv(virtqueue_device_t *queue)
+{
+    virtqueue_ring_object_t handle;
+
+    while (virtqueue_get_available_buf(queue, &handle)) {
+        char temp_buf[VIRTIO_VSOCK_CAMKES_MTU];
```

Comment:
This is tmp buffer is allocated on the stack. Might be nicer to make this part of `virtqueue_device_t`?

Reply:
What's worse it enforces a double copy of the data, in line 101 and in the callback.

Reply:
This is due to the scatter-gather buffer that camkes virtqueue is using. To avoid copying, the emul layer needs to be aware of the structure of the data. Or, I can change the backend to use some other transport mechanisms.